docs/configuration.md
## How to Specify Configurations

The Interledger.rs node binary (`ilp-node`) can be configured in various ways:

### Environment variables

```bash #
# Passing as environment variables
# {parameter name (typically in capital)}={value}
# note that the parameter names MUST begin with a prefix of "ILP_" e.g. ILP_SECRET_SEED
ILP_ADDRESS=example.alice \
ILP_OTHER_PARAMETER=other_value \
ILP_PARENT__CHILD=hierarchical_value \
ilp-node
```

When you want to specify hierarchical parameters such as `bind_address` of `prometheus`, you have to set the parameter name as `ILP_PROMETHEUS__BIND_ADDRESS`, separating the parent and the child with `__` (two underscores). 

### Standard In (stdin)

```bash #
# Passing from STDIN in JSON, TOML, YAML format.
some_command | ilp-node
```

### Configuration files

```bash #
# Passing by a configuration file in JSON, TOML, YAML format.
# The first argument is the path to the configuration file.
ilp-node config.yml
```

### Command line arguments

```bash #
# Passing by command line arguments.
# --{parameter name} {value}
ilp-node --admin_auth_token super-secret --parent.child hierarchical_value
```

When you want to specify hierarchical parameters such as `bind_address` of `prometheus`, you have to set the parameter name as `prometheus.bind_address`, separating the parent and the child with `.` (a dot). 

Note that configurations are applied in the following order of priority:
1. Environment Variables
1. Stdin
1. Configuration files
1. Command line arguments.

## Configuration Parameters

The configuration parameters are explained in the following format.

- name of the config
    - format
    - example
    - explanation

---

### Required

- secret_seed
    - 32 bytes HEX
    - `fe6b34ed652486f38c95e9d761f737cf6473c52b2c8fd3a407fa775ea78e8c82`
    - A secret seed that is used to generate STREAM secrets and used to encrypt sensitive data. This MUST NOT be changed after once you started up the node. You could use `openssl rand -hex 32` to generate one.
- admin_auth_token
    - String
    - `naXg9PrfFAaY99s7`
    - An arbitrary secret token that is used for authenticating against administrative operations over the node's HTTP API. It must be passed as a Bearer token.

### Optional

- ilp_address
    - [ILP Addresses v2.0.0](https://github.com/interledger/rfcs/blob/master/0015-ilp-addresses/0015-ilp-addresses.md)
    - `g.my-node`
    - The ILP address of your node. The format should conform to the RFC above. If you are running a child node, you don't need to specify this.
- database_url
    - URL
    - `redis://127.0.0.1:6379`, `redis+unix:/tmp/redis.sock`
    - A URL of redis that the node connects to in order to store its data.
- http_bind_address
    - Socket Address (`address:port`)
    - `127.0.0.1:7770`
    - A pair of an IP address and a port to listen for HTTP connections. This is used for the HTTP API, ILP over HTTP packets and BTP connections. ILP over HTTP is a means to transfer ILP packets instead of BTP connections.
- settlement_api_bind_address
    - Socket Address (`address:port`)
    - `127.0.0.1:7771`
    - A pair of an IP address and a port to listen for connections from settlement engines. The address provides the Settlement Engine API.
- default_spsp_account
    - String (should be an existing account username)
    - `my_account`
    - When SPSP payments are sent to the root domain, the payment pointer is resolved to `<domain>/.well-known/pay` (if not provided, this endpoint will not be exposed). This value determines which account those payments will be sent to.
- route_broadcast_interval
    - Non-negative Integer (in milliseconds)
    - `30000`
    - Interval, defined in milliseconds, on which the node will broadcast routing information to other nodes using CCP. Defaults to 30000ms (30 seconds).
- exchange_rate
    - provider
        - String (should be one of `CoinCap`, `CryptoCompare`)
        - `CoinCap`
        - Exchange rate API to poll for exchange rates. If this is not set, the node will not poll for rates and will instead use the rates set via the HTTP API. Note that [CryptoCompare](#using-cryptocompare) can also be used **when the node is configured via a config file or stdin**, because an API key must be provided to use that service.
    - poll_interval
        - Non-negative Integer (in milliseconds)
        - `60000`
        - Interval, defined in milliseconds, on which the node will poll the `provider` (if specified) for exchange rates.
    - spread
        - Float
        - `0.01`
        - Spread, as a fraction, to add on top of the exchange rate. This amount is kept as the node operator's profit, or may cover fluctuations in exchange rates. For example, take an incoming packet with an amount of 100. If the exchange rate is 1:0.5 and the spread is 0.01, the amount on the outgoing packet would be 198 (instead of 200 without the spread).
- [prometheus](https://prometheus.io/)
    - bind_address
        - Socket Address (`address:port`)
        - `9654`
        - IP address and port to host the Prometheus exporter on.
    - histogram_window
        - Non-negative Integer (in milliseconds)
        - `300000`
        - Amount of time, in milliseconds, that the node will collect data points for the Prometheus histograms. Defaults to 300000ms (5 minutes).
    - histogram_granularity
        - Non-negative Integer (in milliseconds)
        - `10000`
        - Granularity, in milliseconds, that the node will use to roll off old data. For example, a value of 1000ms (1 second) would mean that the node forgets the oldest 1 second of histogram data points every second. Defaults to 10000ms (10 seconds).

#### Using CryptoCompare 

You have to use a config file or STDIN to use `CryptoCompare` as a rate provider as follows.

```bash #
# run ilp-node with STDIN
some-command | ilp-node
```

Then `some-command` should output a config like:

```yaml
exchange_rate.provider:
  CryptoCompare: insert_api_key_here
```

It is recommended to pass the API key from STDIN because passing from arguments might expose the secret unexpectedly, for example using `history`.


docs/manual-config.md
# Manual Testnet Configuration

## Get Your Own Credentials

In order to connect to the testnet, you need to generate
credentials and get your account configured on a testnet node.

Currently, the [the Xpring testnet website](https://xpring.io/ilp-testnet-creds) can be used to generate credentials. The site automatically configures your account on Xpring's testnet node. You could use the credentials on that node to send and receive payments, or you can use that account's information to connect your local node to the testnet (more on that later)

Just click the **"Generate XRP Credentials"** or **"Generate ETH Credentials"** button, and you'll get your account set on the node. Keep the credentials so that you could later send payments or set up your own node.

The credentials should look like (username and token would be different ones):

|key|value|
|---|---|
|username|user_qjex4cfk|
|token|us5w0zt1br4ft|
|ilp_over_http_url|https://rs3.xpring.dev/ilp|
|ilp_over_btp_url|btp+wss://rs3.xpring.dev/ilp/btp|
|payment_pointer|$rs3.xpring.dev/accounts/user_qjex4cfk/spsp|
|asset_code|XRP|
|asset_scale|9|

You can also obtain the above credentials by making a GET request to `https://xpring.io/api/accounts/xrp` or `https://xpring.io/api/accounts/eth`


## Send and Receive Payments
If you want to just try out the Interledger testnet, you could try as follows.

```bash
curl \
    -X POST \
    -H "Authorization: Bearer ${token}" \
    -H "Content-Type: application/json" \
    -d '{"receiver":"$their-payment-pointer.example","source_amount":500}' \
    https://rs3.xpring.dev/accounts/${Username}/payments
```

Note that you should replace **$their-payment-pointer.example** with one you want to send payments to, and **source_amount** with how much money you want to pay. If you are sending to an account that is configured with a different asset, the returned value may be different from the one you sent, based on any exchange rate calculations.

If someone sends you payments, you could confirm your balance increase as follows.

```bash
curl \
    -H "Authorization: Bearer ${token}" \
    https://rs3.xpring.dev/accounts/${Username}/balance
```

Note: Currently there is no way to check who paid you. This is a feature we intend to support in the future, and can be done by having the sender of the payment include additional metadata in the STREAM packets.



### Receive Payments

Other people could send you payments using the following SPSP address.

```bash
https://<YOUR_NAME>.localtunnel.me/accounts/<YOUR_NAME>/spsp
```

Then you can confirm your balance as follows:

```bash
docker run interledgerrs/ilp-cli --node https://<YOUR_NAME>.localtunnel.me accounts balance <YOUR_NAME> --auth <YOUR_NAME>:<YOUR_NAME>
```

### Manual configuration

**Configuring an Interledger Node is a complicated process. We highly advise you
use the provided docker image which has the Node, Engines and Store packaged
already. Only try doing the next steps if you want to get an in-depth view of
what is happening under the hood**

#### Requirements

Before you try to connect to the testnet, you need to install the following dependencies.

- localtunnel
- Redis
- Rust
- [settlement-xrp](https://github.com/interledgerjs/settlement-xrp) (to settle with XPRL)

1. **localtunnel**

    `localtunnel` exposes your local node to the internet. The others can make connections to your node, which means the Xpring's node would be able to connect to your node.

    ```bash
    npm install -g localtunnel
    ```

1. **Redis**

    The Interledger.rs nodes and settlement engines currently use [Redis](https://redis.io/) to store their data (SQL database support coming soon!). Nodes and settlement engines can use different Redis instances.

    - Compile and install from the source code
    - [Download the source code here](https://redis.io/download)
    - Install using package managers
    - Ubuntu: run `sudo apt-get install redis-server`
    - macOS: If you use Homebrew, run `brew install redis`

    Make sure your Redis is empty. You could run `redis-cli flushall` to clear all the data.

1. **Rust**
    Because Interledger.rs is written in the Rust language, you need the Rust environment. Refer to the [Getting started](https://www.rust-lang.org/learn/get-started) page or just `curl https://sh.rustup.rs -sSf | sh` and follow the instructions.

1. **settlement-xrp**

    Interledger.rs and settlement engines written in other languages are fully interoperable. Here, we'll use the [XRP Ledger Settlement Engine](https://github.com/interledgerjs/settlement-xrp/), which is written in TypeScript. We'll need `node` and `npm` to install and run the settlement engine. If you don't have it already, refer to [Install Node.js](#install-nodejs).

    Install the settlement engine as follows:

    ```bash
    npm install -g ilp-settlement-xrp
    ```

    (This makes the `ilp-settlement-xrp` command available to your PATH.)

### Running a node and peering

If you want to set up your environment manually, you have to:

- Spin up a settlement engine
- Configure and spin up a node
- Set up a localtunnel (in case you don't have your own global address)
- Insert accounts

#### Spin up a Settlement Engine

We currently have settlement engine support for XRP and Ethereum on-ledger Layer-1 transactions (payment channels support is in our pipeline). Set up one depending which asset you chose for your credentials.

To spin up a XRP settlement engine:

```bash
mkdir -p logs
redis-server --port 6380 &> logs/redis_se.log &

DEBUG="settlement* ilp-settlement-xrp" \
CONNECTOR_URL="http://localhost:7771" \
REDIS_PORT=6380 \
ENGINE_PORT=3000 \
ilp-settlement-xrp \
&> logs/settlement-engine-xrpl.log &
```

Note that we have to spin up a Redis which the settlement engine is going to connect to. It automatically get secrets from the XRP faucet, so you don't need to specify. If you want to use your own credentials, you have to provide the environment variable [`XRP_SECRET`](https://github.com/interledgerjs/settlement-xrp/blob/master/src/run.ts#L6). A faucet is available at the [XRPL website](https://xrpl.org/xrp-testnet-faucet.html).

To spin up a Ethereum settlement engine,

```bash
# This should be done outside of the interledger-rs directory, otherwise it will cause an error.
git clone https://github.com/interledger-rs/settlement-engines
cd settlement-engines

mkdir -p logs
redis-server --port 6380 &> logs/redis_se.log &

cargo run --bin ilp-settlement-ethereum -- \
--private_key "${SE_ETH_SECRET}" \
--chain_id 4 \
--confirmations 0 \
--poll_frequency 15000 \
--ethereum_url "${SE_ETH_URL}" \
--connector_url http://127.0.0.1:7771 \
--redis_url redis://127.0.0.1:6380/ \
--settlement_api_bind_address 127.0.0.1:3000 \
&> logs/settlement-engine-eth.log &
```

Note that you have to specify 2 parameters here, `SE_ETH_SECRET` and `SE_ETH_URL`. Because the Xpring node is connected to the Rinkeby testnet of Ethereum, we also need to connect to Rinkeby.

- `SE_ETH_SECRET`
    - The private key to be used. An Ethereum private key is a 32-byte hex string. You can generate one via [Ethereum Generate Wallet](https://github.com/vkobel/ethereum-generate-wallet).
- `SE_ETH_URL`
    - You can either sync an Ethereum node, in which case it will be `http://localhost:8545` or use [Infura](https://infura.io/), in which case the endpoint will look like `rinkeby.infura.io/v3/<API_KEY>`
- Funding your account
    - You can use the [Rinkeby Faucet](https://faucet.rinkeby.io). You will need to make a social media post (Twitter, Facebook, etc.) with your Ethereum address in it. You should then paste the link in the text box, click "Give me Ether" and select the "18.75 ethers / 3 days" option.

#### Configure and Spin up a Node

First, save your config file as `config.json` which contains:

```JSON
{
    "secret_seed": "<secret_seed>",
    "admin_auth_token": "<admin_auth_token>",
    "redis_url": "redis://127.0.0.1:6379/",
    "http_bind_address": "127.0.0.1:7770",
    "settlement_api_bind_address": "127.0.0.1:7771"
}
```

You have to replace `<secret_seed>` with your secret seed which is used to encrypt information. You could generate one with `openssl rand -hex 32`. `<admin_auth_token>` should be replaced with your admin token, which would be used when you make API calls to the admin-only endpoints of the node.

Then try:

```bash
redis-server --port 6379 &> logs/redis.log &
cargo run --all-features --bin ilp-node -- config.json &> logs/node.log &
```

Now you have your own node running locally🎉

#### Set up a `localtunnel`

In most cases, you will not have a global address for your node. In that case, you could utilize [`localtunnel`](http://localtunnel.me) to make the other nodes connect to your node without any global addresses. Try:

```bash
lt -p 7770 -s <YOUR_NAME> &>logs/localtunnel.log &
```

You can replace `YOUR_NAME` with any name of your choice. Your node will then be reachable via `<YOUR_NAME>.localtunnel.me`.

#### Insert Accounts

Now that our node is running and exposed to the internet, let's insert accounts. We will use the `ilp-cli` binary to insert the following:

1. Default settlement engines endpoints
    ```
    cargo run --bin ilp-cli settlement-engines set-all \
        --auth=<admin_auth_token> \
        --pair XRP http://localhost:3001 \
        --pair ETH http://localhost:3002
    ```

1. The main account that we use for sending and receiving payments.
    ```
    cargo run --bin ilp-cli accounts create <YOUR_NAME> \
        --auth=<admin_auth_token> \
        --asset-code=<ASSET_CODE> \
        --asset-scale=<ASSET_SCALE>
        --ilp-over-http-incoming-token=<YOUR_PASSWORD>
    ```

1. Connect to the Xpring Testnet node so we're connected to the rest of the testnet
    ```
    cargo run --bin ilp-cli testnet setup <ASSET_CODE> \
         --auth=<admin_auth_token> \
         --return-testnet-credential
    ```
    The above command will return the username and password, in the form of `USERNAME:PASSWORD` which we use to modify our account's settings the testnet node.

1. (Optional) Configure the testnet node to talk to us over HTTP instead of BTP (HTTP is more reliable, BTP is more efficient due to WebSockets)
   1. Configure the testnet account on our node to authenticate with ILP over HTTP
       ```
        cargo run --bin ilp-cli accounts update-settings xpring_<ASSET_CODE> \
                --auth=<admin_auth_token> \
                --ilp-over-http-incoming-token=xpring_password \
                --settle-threshold=1000 \
                --settle-to=0
        ```

   1.  Configure ILP over HTTP for our account on the testnet node
        ```
        cargo run --bin ilp-cli --node=https://rs3.xpring.dev accounts update-settings $USERNAME
                --auth=$TESTNET_AUTH \
                --ilp-over-http-outgoing-token=xpring_<ASSET_CODE>:xpring_password \
                --ilp-over-http-url=https://<YOUR_NAME>.localtunnel.me/ilp \
                --settle-threshold=1000 \
                --settle-to=0
        ```

    Notes:
        - `$USERNAME` is the first part of the returned value from the previous step.
        - `$TESTNET_AUTH` is the full `USERNAME:PASSWORD` string from the previous step


The peering process with the testnet is done! Congratulations!

You should now be able to send and receive payments through the testnet node.


settlement-engines/xrp/src/cli.ts
#!/usr/bin/env node

import * as yargs from 'yargs'
import { XrpSettlementEngine } from './index'

const testnetServer = 'wss://s.altnet.rippletest.net:51233'

const argv = yargs.option('address', {
    description: 'XRP address to settle from and listen to for incoming payments',
    type: 'string'
}).option('secret', {
    description: 'XRP account secret',
    type: 'string'
}).option('testnet', {
    default: false,
    type: 'boolean',
    description: 'Connect to the XRP testnet instead of the main net'
}).option('rippled', {
    description: 'Websocket URI of the rippled server to connect to',
    default: 'wss://s1.ripple.com',
    type: 'string'
}).option('redis', {
    description: 'Redis URI to connect to',
    // TODO maybe connect using a socket instead
    default: 'redis://localhost:6379',
    type: 'string'
}).option('min_settlement_amount', {
    default: 1000000,
    type: 'number',
    description: 'Minimum amount, denominated in XRP drops (0.000001 XRP), to send in an XRP payment'
}).option('poll_interval', {
    default: 60000,
    type: 'number',
    description: 'Interval, denominated in milliseconds, to poll the Redis database for changes in account balances'
}).require(['address', 'secret'])
    .argv

const config = {
  address: argv.address,
  secret: argv.secret,
  rippledUri: (argv.testnet ? testnetServer : argv.rippled),
  redisUri: argv.redis,
  minSettlementAmount: argv.min_settlement_amount,
  pollInterval: argv.poll_interval
}

const engine = new XrpSettlementEngine(config)
engine.connect().then(() => {
  console.log('Listening for incoming XRP payments and polling Redis for accounts that need to be settled')
}).catch((err) => console.error(err))
